### PR TITLE
Re-add icon=url

### LINKED ISSUE #19: Re-add icon=url
Import from https://bugzilla.wikimedia.org/show_bug.cgi?id=62061

In Maps 2 you could use icon=/w/images/c/ce/Green_marker.png or icon=//semantic-mediawiki.org/w/images/c/ce/Green_marker.png. Maps 3 lost that possibility, you can only use icon= Green_marker.png . Can this be readded?

--  egel@scoutnet.nl 2014-02-28 15:37:35 UTC

### PR DESCRIPTION
**Pull request by PeterTheOne**
fix #19 

untested.

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
This at the very least needs a manual test, and preferably an automated one (which should not be too hard).

**Comment by PeterTheOne** [comment]
I did a manual test that worked and I wrote a phpunit test but only tested for the case that no imagePage exists.

File: tests/phpunit/MapsMapperTest.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace Maps\Test;
+
+use MapsMapper;
+
+/**
+ * @covers MapsMapper
```
Comment (PeterTheOne) [inline tests/phpunit/MapsMapperTest.php]: Is this correct since the test class doesn't cover all of MapsMapper?

Comment (JeroenDeDauw) [inline tests/phpunit/MapsMapperTest.php]: It is... what this does is avoid counting coverage of lines outside that class.

File: tests/phpunit/MapsMapperTest.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * @since 3.6
+ *
+ * @group Maps
+ *
+ * @licence GNU GPL v2+
+ * @author Peter Grassberger < petertheone@gmail.com >
+ */
+class MapsMapperTest extends \MediaWikiTestCase {
+
+    public static $imageUrls = [
```
Comment (JeroenDeDauw) [inline tests/phpunit/MapsMapperTest.php]: Does this need to be public? And does it need to be static?

File: tests/phpunit/MapsMapperTest.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ * @covers MapsMapper
+ *
+ * @since 3.6
+ *
+ * @group Maps
+ *
+ * @licence GNU GPL v2+
+ * @author Peter Grassberger < petertheone@gmail.com >
+ */
+class MapsMapperTest extends \MediaWikiTestCase {
```
Comment (JeroenDeDauw) [inline tests/phpunit/MapsMapperTest.php]: Can't you just derive from the base PHPUnit test class?

File: tests/phpunit/MapsMapperTest.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    public static $imageUrls = [
+        'markerImage.png' => 'markerImage.png',
+        '/w/images/c/ce/Green_marker.png' => '/w/images/c/ce/Green_marker.png',
+        '//semantic-mediawiki.org/w/images/c/ce/Green_marker.png' => '//semantic-mediawiki.org/w/images/c/ce/Green_marker.png',
+        'Cat2.jpg' => 'Cat2.jpg'
+    ];
+
+    function getImagePage( $filename ) {
```
Comment (JeroenDeDauw) [inline tests/phpunit/MapsMapperTest.php]: private?

File: tests/phpunit/MapsMapperTest.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        $iPage->setFile( $file );
+        return $iPage;
+    }
+
+    /**
+     * Tests MapsMapperTest::getFileUrl()
+     */
+    public function testGetFileUrl() {
+        foreach ( self::$imageUrls as $rawValue => $parsedValue ) {
+            $this->assertEquals( $parsedValue, MapsMapper::getFileUrl($rawValue) );
```
Comment (JeroenDeDauw) [inline tests/phpunit/MapsMapperTest.php]: If you compare strings, use `assertSame`

File: tests/phpunit/MapsMapperTest.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        }
+        $this->assertEquals( null, MapsMapper::getFileUrl(null));
```
Comment (JeroenDeDauw) [inline tests/phpunit/MapsMapperTest.php]: `assertNull`

This belongs in a separate test method: `testGivenNull_getFileUrlReturnsNull`

File: tests/phpunit/MapsMapperTest.php [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        // TODO test with existing imagePage
```
Comment (JeroenDeDauw) [inline tests/phpunit/MapsMapperTest.php]: That would also go in a different method, so TODO should be on class level

File: tests/phpunit/MapsMapperTest.php [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+        $iPage = new ImagePage( $title );
+        $iPage->setFile( $file );
+        return $iPage;
+    }
+
+    /**
+     * Tests MapsMapperTest::getFileUrl()
+     */
+    public function testGetFileUrl() {
+        foreach ( self::$imageUrls as $rawValue => $parsedValue ) {
```
Comment (JeroenDeDauw) [inline tests/phpunit/MapsMapperTest.php]: https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers

**Comment by JeroenDeDauw** [comment]
Thank you for writing the test! Some minor things to take care off tho

File: tests/phpunit/MapsMapperTest.php [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        return [
+            ['markerImage.png', 'markerImage.png'],
+            ['/w/images/c/ce/Green_marker.png', '/w/images/c/ce/Green_marker.png'],
+            ['//semantic-mediawiki.org/w/images/c/ce/Green_marker.png', '//semantic-mediawiki.org/w/images/c/ce/Green_marker.png'],
+            ['Cat2.jpg', 'Cat2.jpg'],
+        ];
+    }
+
+    /**
+     * Tests MapsMapperTest::getFileUrl()
```
Comment (JeroenDeDauw) [inline tests/phpunit/MapsMapperTest.php]: I think that part is superfluous. And when using it, I'd use a @covers tag as well, so its easier to detect when its wrong. Not important tho

**Pull request merged**